src/hotspot/share/oops/klass.inline.hpp

```diff
@@ -76,4 +77,60 @@ inline ByteSize Klass::vtable_start_offset() {
   return in_ByteSize(InstanceKlass::header_size() * wordSize);
 }
 
+inline bool Klass::lookup_secondary_supers_table(Klass* k) const {
+  uintx bitmap = _bitmap;
+
+  if (bitmap == SECONDARY_SUPERS_BITMAP_FULL) {
+    return linear_search_secondary_supers(k);
+  }
+
+  constexpr int highest_bit_number = SECONDARY_SUPERS_TABLE_SIZE - 1;
+  uint8_t slot = k->_hash_slot;
+  uintx shifted_bitmap = bitmap << (highest_bit_number - slot);
+
+  // First check the bitmap to see if super_klass might be present. If
+  // the bit is zero, we are certain that super_klass is not one of
+  // the secondary supers.
+  if ((shifted_bitmap >> highest_bit_number) == 0) {
+    return false;
+  }
+
+  // Calculate the initial hash probe
+  int index = population_count(shifted_bitmap) - 1;
+  if (secondary_supers()->at(index) == k) {
+    // Yes! It worked the first time.
+    return true;
+  }
+
+  // Is there another entry to check? Consult the bitmap. If Bit 1,
+  // the next bit to test, is zero, we are certain that super_klass is
+  // not one of the secondary supers.
+  bitmap = rotate_right(bitmap, slot);
+  if ((bitmap & 2) == 0) {
+    return false;
+  }
+
+  // Continue probing the hash table
+  return fallback_search_secondary_supers(k, index, bitmap);
+}
+
+inline bool Klass::search_secondary_supers(Klass *k) const {
+  // This is necessary because I am never in my own secondary_super list.
+  if (this == k)
+    return true;
+
+  bool result = lookup_secondary_supers_table(k);
```

Comment:
I think not. It'd complicate C++ runtime for no useful reason.

Reply:
On the other hand, if `-XX:-UseSecondarySupersTable` is intended solely for diagnostic purposes, then handling all possible execution modes uniformly is preferable, since it gives more confidence when troubleshooting seemingly related failures.

Reply:
Alternatively, `Klass::is_subtype_of()` can unconditionally perform linear search over secondary_supers array. 

Even though I very much like to see table lookup written in C++ (accompanying heavily optimized platform-specific MacroAssembler variants), it would make C++ runtime even simpler.

Reply:
> Also also, Klass::is_subtype_of() is used for C1 runtime.

Can you elaborate, please? What I'm seeing in `Runtime1::generate_code_for()` for `slow_subtype_check` is a call into `MacroAssembler::check_klass_subtype_slow_path()`.

Reply:
> > Also also, Klass::is_subtype_of() is used for C1 runtime.
> 
> Can you elaborate, please?

Sorry, that was rather vague. In C1-compiled code, the Java method `Class::isInstance(Object)`calls `Klass::is_subtype_of()`. 

In general, I find it difficult to decide how much work, if any, should be done to improve C1 performance. Clearly, if C1 exists only to help with startup time in a tiered compilation system, the answer is "not much".

Reply:
Thanks, now I see that `Class::isInstance(Object)` is backed by `Runtime1::is_instance_of()` which uses `oopDesc::is_a()` to do the job.

If it turns out to be performance critical, the intrinsic implementation should be rewritten to exercise existing subtype checking support in C1. As it is implemented now, it's already quite inefficient.

Reply:
I did write an intrinsic for that, but it made this patch even larger. I have a small patch for C1, for some other time.

Reply:
FYI I filed a low-priority RFE against C1 to track it.